src/lib/encryption/hkdf-encryption-protocol.ts

```diff
@@ -0,0 +1,331 @@
+import crypto from "crypto";
+import { z } from "zod";
+import { Err, Ok, Result } from "../utilities/result";
+import { EncryptionProtocol } from "./encryption-protocol";
+
+export const HkdfInputKeyMaterialSchema = z.object({
+  ikmId: z
+    .string()
+    .min(1)
+    .max(6)
+    .refine((val) => {
+      const parts = val.split(".");
+      return parts.length === 1;
+    }),
+  ikmHex: z.string(),
+});
+
+export type HkdfInputKeyMaterial = z.infer<typeof HkdfInputKeyMaterialSchema>;
+
+/**
+ * Configuration for the protocol.
+ *
+ * "ikms" is an array of IKMs. When encrypting, the IKM at index 0 will be used.
+ *
+ * "purpose" is incorporated into the "info" during the key derivation step.
+ */
+export const HkdfConfigSchema = z.object({
+  ikms: z.array(HkdfInputKeyMaterialSchema).min(1),
+  purpose: z.string(),
+});
+
+export type HkdfConfig = z.infer<typeof HkdfConfigSchema>;
+
+const _ENC2_ = "_ENC2_";
+
+export class HkdfEncryptionProtocol implements EncryptionProtocol {
+  private ikms: Record<string, Buffer>;
+  private purpose: string;
+  private ikmId: string;
+
+  constructor(config: HkdfConfig) {
+    const { ikms, purpose } = HkdfConfigSchema.parse(config);
+    this.purpose = purpose;
+    this.ikmId = ikms[0].ikmId;
+    this.ikms = {};
+    for (const ikm of ikms) {
+      this.ikms[ikm.ikmId] = Buffer.from(ikm.ikmHex, "hex");
+    }
+  }
+
+  name(): string {
+    return "HkdfEncryptionProtocol";
+  }
+
+  async encrypt(data: string): Promise<Result<{ encrypted: string }, string>> {
+    const { ikms, purpose, ikmId } = this;
+    const { hkdfArgs } = _generateHkdfArgs({ ikms, ikmId, purpose });
+    const { key: encKey } = await _deriveKey({ hkdfArgs, usage: "enc" });
+    const { key: sigKey } = await _deriveKey({ hkdfArgs, usage: "sig" });
+    const { cipherArgs } = _generateCipherArgs({ encKey: encKey });
+    const { ciphertext } = _encrypt({ data, cipherArgs });
+    const { hkdfParams } = _toHkdfParams({ hkdfArgs });
+    const { cipherParams } = _toCipherParams({ cipherArgs });
+    const { payload } = _toPayload({ hkdfParams, cipherParams, ciphertext });
+    const { signature } = _sign({ sigKey, payload });
+    const { encrypted } = _pack({ ikmId, signature, payload });
+    return Ok({ encrypted });
+  }
+
+  async decrypt(encrypted: string): Promise<Result<{ data: string }, string>> {
+    const { ikms } = this;
+
+    const { ikmId, signature, payload } = _unpack({ encrypted });
+    const { hkdfParams, cipherParams, ciphertext } = _fromPayload({ payload });
+
+    const resHkdfArgs = _toHkdfArgs({ ikms, hkdfParams });
+    if (resHkdfArgs.error !== undefined) {
+      return Err(resHkdfArgs.error);
+    }
```

Comment:
I think if we leave the keys out of the Args, and instead pass them into _encrypt, _decrypt, and _sign explicitly, it will be clearer where the keys are used, and also we won't have this ugly part. Also it will be clearer that params do not contain keys.

Reply:
Done in 4367fd8